### DISCUSSION TITLE
NLOHMANN_DEFINE_TYPE_* does not provide definitions for ordered_json

### DISCUSSION THREAD

**Discussion by MHebes** [nlohmann/json#2397] [Ideas]
#### What is the issue you have?

When using the macros suggested [here](https://nlohmann.github.io/json/features/arbitrary_types/#simplify-your-life-with-macros) to automatically create serialization methods for structs, to_json is only defined with a concrete `json` type.

#### Please describe the steps to reproduce the issue.

1. Create a struct.
2. Add to_json and from_json methods automatically using the macro.
3. Attempt to initialize a ordered_json with the struct.

#### Can you provide a small but working code example?

```c++
#include "json.hpp"

typedef struct struct_t {
  int c = 1;
  int b = 2;
  int a = 3;

  NLOHMANN_DEFINE_TYPE_INTRUSIVE(struct_t, c, b, a);
} struct_t;

int main() {
  struct_t str;

  nlohmann::json j = str; // fine
  // nlohmann::ordered_json j = str; // error: not assignable

  std::cout << j.dump(4) << std::endl;
}
```

#### What is the expected behavior?

The assignment to the ordered_json should behave in the same way as assignment to json.

#### And what is the actual behavior instead?

The code snippet does not compile when the commented line is used.

#### Workaround

Define to_json and from_json manually:

```c++
  friend void to_json(nlohmann::ordered_json &j, const struct_t &t) {
    j["c"] = t.c;
    j["b"] = t.b;
    j["a"] = t.a;
  }
```

Or use this macro instead:

```c++
#define AS_JSON(Type, ...)                                                     \
  friend void to_json(nlohmann::ordered_json &nlohmann_json_j,                 \
                      const Type &nlohmann_json_t) {                           \
    NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, __VA_ARGS__))   \
  }                                                                            \
  friend void from_json(const nlohmann::ordered_json &nlohmann_json_j,         \
                        Type &nlohmann_json_t) {                               \
    NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_FROM, __VA_ARGS__)) \
  }
```

#### Suggested fix

Either add ordered_json overloads in the macro itself, or possibly do some template magic to use basic_json in the parameter list.

#### Less ideal fix

Add a new macro specific to ordered_json

#### Which compiler and operating system are you using?

- Compiler: Microsoft Visual C++ 2019
- Operating system: Windows

#### Which version of the library did you use?

- [x] latest release version 3.9.1
- [ ] other release - please state the version: ___
- [ ] the `develop` branch

P.S. Love the library :)

**Comment by ScottHutchinson** [comment]
I had the same issue, and adding this macro seems to fix it.
```cpp
/*!
@brief macro added to work with ordered_json -- hutch
@def NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_UNORDERED
*/
#define NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_UNORDERED(Type, ...)  \
    inline void to_json(nlohmann::ordered_json& nlohmann_json_j, const Type& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, __VA_ARGS__)) } \
    inline void from_json(const nlohmann::ordered_json& nlohmann_json_j, Type& nlohmann_json_t) { NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_FROM, __VA_ARGS__)) }
```

**Reply by nlohmann** [reply]
Thanks for sharing - this should also be fixed with https://github.com/nlohmann/json/pull/2532.
